### ISSUE TITLE
Overuse of "local:" when reading SBML files

### ISSUE THREAD

**Issue by jhgennari** [sys-bio/libOmexMeta#48]
The RDF.from_file() inappropriately uses the "local:" namespace when reading annotations from SBML files. This can be easily seen with the species annotations from BioModels #385. For example, species #32 is identified as bqbiol:isVersionof CHEBI:16526. The system currently creates: 
local:metaid_0000032
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .

Whereas it should be 
<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000032>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .

The "local:" namespace refers to NewModel.rdf, not the SBML file of NewModel.xml. 

The preferred output would actually be 
<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000032>
    bqbiol:isVersionOf <http://identifiers.org/obo.chebi/CHEBI:16526> . 

but that would require dealing (correctly) with the rdf:Bag construct.

**Comment by CiaranWelsh** [comment]
Minimal example of John's issue (no new information). 

```
<?xml version="1.0" encoding="UTF-8"?>
<sbml xmlns="http://www.sbml.org/sbml/level3/version1/core" level="3" version="1">
    <model metaid="__main" id="__main">
        <listOfCompartments>
            <compartment sboTerm="SBO:0000410" id="default_compartment" spatialDimensions="3" size="1" constant="true"/>
        </listOfCompartments>
        <listOfSpecies>
            <species id="S1" metaid="S1" compartment="default_compartment" initialConcentration="1" hasOnlySubstanceUnits="false"
                     boundaryCondition="false" constant="false">
                <annotation>
                    <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
                             xmlns:bqbiol="http://biomodels.net/biology-qualifiers/">
                        <rdf:Description rdf:about="#S1">
                            <bqbiol:isVersionOf>
                                <rdf:Bag>
                                    <rdf:li rdf:resource="http://identifiers.org/obo.chebi/CHEBI:16526"/>
                                </rdf:Bag>
                            </bqbiol:isVersionOf>
                        </rdf:Description>
                    </rdf:RDF>
                </annotation>
            </species>
            <species id="S2" compartment="default_compartment" initialConcentration="3" hasOnlySubstanceUnits="false"
                     boundaryCondition="false" constant="false"/>
        </listOfSpecies>
        <listOfParameters>
            <parameter id="k1" value="0.1" constant="true"/>
        </listOfParameters>
        <listOfReactions>
            <reaction id="_J0" reversible="false" fast="false">
                <listOfReactants>
                    <speciesReference species="S1" stoichiometry="1" constant="true"/>
                </listOfReactants>
                <listOfProducts>
                    <speciesReference species="S2" stoichiometry="1" constant="true"/>
                </listOfProducts>
                <kineticLaw>
                    <math xmlns="http://www.w3.org/1998/Math/MathML">
                        <apply>
                            <times/>
                            <ci>k1</ci>
                            <ci>S1</ci>
                        </apply>
                    </math>
                </kineticLaw>
            </reaction>
        </listOfReactions>
    </model>
</sbml>
```

Is read by `RDF::fromString` as
```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix bqbiol: <http://biomodels.net/biology-qualifiers/> .
@prefix semsim: <http://bime.uw.edu/semsim/> .
@prefix OMEXlib: <http://omex-library.org/> .
@prefix local: <http://omex-library.org/NewOmex.omex/NewModel.rdf#> .

local:Process0000
    semsim:hasSinkParticipant local:SinkParticipant0000 ;
    semsim:hasSourceParticipant local:SourceParticipant0000 .

local:S1
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .

local:SinkParticipant0000
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#species0000> .

local:SourceParticipant0000
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#S1> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#reaction0000>
    bqbiol:isPropertyOf local:Process0000 ;
    bqbiol:isVersionOf <https://identifiers.org/opb:OPB_00592> .
```

note that  
```
http://omex-library.org/NewOmex.omex/NewModel.xml#S1
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .
```
Should be 
```
local:S1
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .
```

Or better: 

```
http://omex-library.org/NewOmex.omex/NewModel.xml#S1 bqbiol:isVersionOf http://identifiers.org/obo.chebi/CHEBI:16526 .
```

**Comment by CiaranWelsh** [comment]
So, here's a snippet from the code that does the parsing: 

```
    RDF RDF::fromString(const std::string &str, const std::string &format) {
        RDF rdf;
        LibrdfParser parser(format);
        *LibrdfUri u(rdf.getLocalUri());*
        parser.parseString(str, rdf.model_, u);
        u.freeUri();
       ...
```
By changing the url that is used for parsing from `*LibrdfUri u(rdf.getLocalUri());*` to `*LibrdfUri u(rdf.getModelUri());*`

aka

```
    RDF RDF::fromString(const std::string &str, const std::string &format) {
        RDF rdf;
        LibrdfParser parser(format);
        *LibrdfUri u(rdf.getModelUri());*
        parser.parseString(str, rdf.model_, u);
        u.freeUri();
       ...
```

We get the desired effect: 

```

@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix bqbiol: <http://biomodels.net/biology-qualifiers/> .
@prefix semsim: <http://bime.uw.edu/semsim/> .
@prefix OMEXlib: <http://omex-library.org/> .
@prefix local: <http://omex-library.org/NewOmex.omex/NewModel.rdf#> .

local:Process0000
    semsim:hasSinkParticipant local:SinkParticipant0000 ;
    semsim:hasSourceParticipant local:SourceParticipant0000 .

local:SinkParticipant0000
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#species0000> .

local:SourceParticipant0000
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#S1> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#S1>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#reaction0000>
    bqbiol:isPropertyOf local:Process0000 ;
    bqbiol:isVersionOf <https://identifiers.org/opb:OPB_00592> .

```
However I'm uncertain of any potentially undesirable side effects, since the `model uri` will be used for all parsing (note: this occurs before the sbml extraction stuff, which is uneffected by this change)

**Comment by jhgennari** [comment]
Ciaran -- probably the best way to test would be to run the modified code on BioModels #385, and then send the results to me. That way we can at least spot check for side effects.

**Comment by CiaranWelsh** [comment]
With the changes above, which are implemented in #51, the RDF produced by libOmexMeta for BIOMD385 is pasted below. 

```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix bqmodel: <http://biomodels.net/model-qualifiers/> .
@prefix bqbiol: <http://biomodels.net/biology-qualifiers/> .
@prefix semsim: <http://bime.uw.edu/semsim/> .
@prefix OMEXlib: <http://omex-library.org/> .
@prefix local: <http://omex-library.org/NewOmex.omex/NewModel.rdf#> .

local:MediatorParticipant0000
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000033> .

local:MediatorParticipant0001
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000032> .

local:MediatorParticipant0002
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000031> .

local:Process0000
    semsim:hasMediatorParticipant local:MediatorParticipant0000 ;
    semsim:hasSinkParticipant local:SinkParticipant0000 ;
    semsim:hasSourceParticipant local:SourceParticipant0000, local:SourceParticipant0001, local:SourceParticipant0002 .

local:Process0001
    semsim:hasMediatorParticipant local:MediatorParticipant0001 ;
    semsim:hasSinkParticipant local:SinkParticipant0001 ;
    semsim:hasSourceParticipant local:SourceParticipant0003, local:SourceParticipant0004, local:SourceParticipant0005 .

local:Process0002
    semsim:hasMediatorParticipant local:MediatorParticipant0002 ;
    semsim:hasSinkParticipant local:SinkParticipant0002 ;
    semsim:hasSourceParticipant local:SourceParticipant0006 .

local:Process0003
    semsim:hasSinkParticipant local:SinkParticipant0003 ;
    semsim:hasSourceParticipant local:SourceParticipant0007 .

local:SinkParticipant0000
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000030> .

local:SinkParticipant0001
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000030> .

local:SinkParticipant0002
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000029> .

local:SinkParticipant0003
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000031> .

local:SourceParticipant0000
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000029> .

local:SourceParticipant0001
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000032> .

local:SourceParticipant0002
    semsim:hasMultiplier "2"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000031> .

local:SourceParticipant0003
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000029> .

local:SourceParticipant0004
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000033> .

local:SourceParticipant0005
    semsim:hasMultiplier "2"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000031> .

local:SourceParticipant0006
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000030> .

local:SourceParticipant0007
    semsim:hasMultiplier "1"^^rdf:double ;
    semsim:hasPhysicalEntityReference <http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000034> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#_272044>
    bqbiol:hasTaxon [
        rdf:_1 <http://identifiers.org/taxonomy/33090> ;
        a rdf:Bag
    ] ;
    bqmodel:is [
        rdf:_1 <http://identifiers.org/biomodels.db/MODEL1109270001> ;
        a rdf:Bag
    ], [
        rdf:_1 <http://identifiers.org/biomodels.db/BIOMD0000000385> ;
        a rdf:Bag
    ], [
        rdf:_1 <http://identifiers.org/obo.go/GO:0019253> ;
        a rdf:Bag
    ] ;
    bqmodel:isDescribedBy [
        rdf:_1 <http://identifiers.org/pubmed/22001849> ;
        a rdf:Bag
    ] ;
    <http://purl.org/dc/elements/1.1/creator> [
        rdf:_1 [
            <http://www.w3.org/2001/vcard-rdf/3.0#EMAIL> "viji@ebi.ac.uk" ;
            <http://www.w3.org/2001/vcard-rdf/3.0#N> [
                <http://www.w3.org/2001/vcard-rdf/3.0#Family> "Chelliah" ;
                <http://www.w3.org/2001/vcard-rdf/3.0#Given> "Vijayalakshmi"
            ] ;
            <http://www.w3.org/2001/vcard-rdf/3.0#ORG> [
                <http://www.w3.org/2001/vcard-rdf/3.0#Orgname> "EMBL-EBI"
            ]
        ] ;
        rdf:_2 [
            <http://www.w3.org/2001/vcard-rdf/3.0#EMAIL> "nikoloski@mpimp-golm.mpg.de" ;
            <http://www.w3.org/2001/vcard-rdf/3.0#N> [
                <http://www.w3.org/2001/vcard-rdf/3.0#Family> "Nikoloski" ;
                <http://www.w3.org/2001/vcard-rdf/3.0#Given> "Zoran"
            ] ;
            <http://www.w3.org/2001/vcard-rdf/3.0#ORG> [
                <http://www.w3.org/2001/vcard-rdf/3.0#Orgname> "Institute  of Biochemistry and Biology, University of Potsdam, 14476 Potsdam, Germany"
            ]
        ] ;
        rdf:_3 [
            <http://www.w3.org/2001/vcard-rdf/3.0#EMAIL> "arnold@mpimp-golm.mpg.de" ;
            <http://www.w3.org/2001/vcard-rdf/3.0#N> [
                <http://www.w3.org/2001/vcard-rdf/3.0#Family> "Arnold" ;
                <http://www.w3.org/2001/vcard-rdf/3.0#Given> "Anne"
            ] ;
            <http://www.w3.org/2001/vcard-rdf/3.0#ORG> [
                <http://www.w3.org/2001/vcard-rdf/3.0#Orgname> "Max-Planck-Institute of Molecular Plant Physiology"
            ]
        ] ;
        a rdf:Bag
    ] ;
    <http://purl.org/dc/terms/created> [
        <http://purl.org/dc/terms/W3CDTF> "2011-10-19T14:51:13Z"
    ] ;
    <http://purl.org/dc/terms/modified> [
        <http://purl.org/dc/terms/W3CDTF> "2012-04-20T19:52:45Z"
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000029>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16710> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000030>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:17050> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000031>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16474> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000032>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:16526> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000033>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:15379> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000034>
    bqbiol:isVersionOf [
        rdf:_1 <http://identifiers.org/obo.chebi/CHEBI:25523> ;
        a rdf:Bag
    ] .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000042>
    bqbiol:isPropertyOf local:Process0000 ;
    bqbiol:isVersionOf <https://identifiers.org/opb:OPB_00592> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000043>
    bqbiol:isPropertyOf local:Process0001 ;
    bqbiol:isVersionOf <https://identifiers.org/opb:OPB_00592> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000044>
    bqbiol:isPropertyOf local:Process0002 ;
    bqbiol:isVersionOf <https://identifiers.org/opb:OPB_00592> .

<http://omex-library.org/NewOmex.omex/NewModel.xml#metaid_0000045>
    bqbiol:isPropertyOf local:Process0003 ;
    bqbiol:isVersionOf <https://identifiers.org/opb:OPB_00592> .

```

**Comment by CiaranWelsh** [comment]
implemented in #47
